system/jlib/jptree.cpp

```diff
@@ -8279,3 +8293,291 @@ IPropertyTree *createPTreeFromYAMLFile(const char *filename, byte flags, PTreeRe
+    }
+
+    Owned<IAttributeIterator> it = tree->getAttributes(true);
+    bool hasAttributes = it->first();
+    bool complex = (hasAttributes || tree->hasChildren());
+    if (complex)
+        yaml.beginMap();
+
+    if (hasAttributes)
+    {
+        ForEach(*it)
+        {
+            const char *key = it->queryName()+1;
+            const char *val = it->queryValue();
+            yaml.writeAttribute(key, val, checkYAMLHide(val, flags, true, false));
+        }
+    }
+    StringBuffer _content;
+    const char *content = nullptr; // to avoid uninitialized warning
+    bool isBinary = tree->isBinary(NULL);
+    bool isNull;
+    if (isBinary)
+    {
+        MemoryBuffer thislevelbin;
+        isNull = (!tree->getPropBin(NULL, thislevelbin))||(thislevelbin.length()==0);
+        if (!isNull)
+            JBASE64_Encode(thislevelbin.toByteArray(), thislevelbin.length(), _content, true);
+        content = _content.str();
+    }
+    else if (tree->isCompressed(NULL))
+    {
+        isNull = false; // can't be empty if compressed;
+        verifyex(tree->getProp(NULL, _content));
+        content = _content.str();
+    }
+    else
+        isNull = (NULL == (content = tree->queryProp(NULL)));
+
+    if (isNull && !root && !complex)
+    {
+        yaml.writeValue("null", false, false, false);
+        return;
+    }
+
+    Owned<IPropertyTreeIterator> sub = tree->getElements("*", 0 != (flags & YAML_SortTags) ? iptiter_sort : iptiter_null);
+    //note that detection of repeating elements relies on the fact that ptree elements
+    //of the same name will be grouped together
+    bool repeatingElement = false;
+    sub->first();
+    while(sub->isValid())
+    {
+        Linked<IPropertyTree> element = &sub->query();
+        const char *name = element->queryName();
+        if (sub->next() && !repeatingElement && streq(name, sub->query().queryName()))
+        {
+            yaml.beginSequence(name);
+            repeatingElement = true;
+        }
+
+        _toYAML(element, yaml, flags, false, repeatingElement);
+
+        if (repeatingElement && (!sub->isValid() || !streq(name, sub->query().queryName())))
+        {
+            yaml.endSequence();
+            repeatingElement = false;
+        }
+    }
+
+    if (!isNull)
+    {
+        if (complex)
+            yaml.writeName("^");
+        //repeating/array/sequence items are implicitly elements, no need for tag
+        yaml.writeValue(content, isArrayItem ? false : true, checkYAMLHide(content, flags, false, isBinary), isBinary);
+    }
+
+    if (complex)
+        yaml.endMap();
+}
+
+
+static void _toYAML(const IPropertyTree *tree, IIOStream &out, unsigned indent, byte flags, bool root=false, bool isArrayItem=false)
+{
+    YAMLEmitter yaml(out, indent);
+    _toYAML(tree, yaml, flags, true, false);
+}
+
+jlib_decl StringBuffer &toYAML(const IPropertyTree *tree, StringBuffer &ret, unsigned indent, byte flags)
+{
+    class CAdapter : implements IIOStream, public CInterface
```

Comment:
minor: This could use CInterfaceOf<IIoStream>  It could also be a single common class instead of being duplicated a few times in this file.